### PR TITLE
VReplication/OnlineDDL: reordering enum values

### PR DESCRIPTION
**Pull request by shlomi-noach**
## Description

Fixes https://github.com/vitessio/vitess/issues/15349

This PR builds a test case where an `enum`'s values are reordered, from `('red', 'green', 'blue')` to `('blue', 'green', 'red')`, and observes how `vreplication` (in this case via OnlineDDL) handles it.

The behavior before this PR (see test output) is that `vcopier` will copy enum values by text, and `vplayer` by ordinal value.

As of this PR, and solved specifically for Online DDL (the problem is irrelevant for other `vreplication` workflows), `vplayer` applies `enum` column changes by writing the _logical_ (string representation, e.g. `'blue'`) of the column.

The change is actually quite small. A prior work on converting `enum`s to textual columns has already laid the necessary foundation: making `vplayer` aware of the enum's schema. This is already supported by setting `vreplication`'s `Filter.Rule.ConvertEnumToText` map. Previously, we only populated it, as mentioned, when converting `enum` to text. The only change needed in this PR was to populate it also when an `enum` remains an `enum`. Whether the `enum`'s schema is change or not, this ensures applying binlog events from the original table to the target table by enum textual value.

### Backport 

As this is a corruption bugfix, I believe this should be backported to all supported versions.

## Related Issue(s)

- https://github.com/vitessio/vitess/issues/15349
- https://github.com/vitessio/vitess/issues/15105
- https://github.com/vitessio/vitess/pull/15106

## Checklist

-   [ ] "Backport to:" labels have been added if this change should be back-ported to release branches
-   [ ] If this change is to be back-ported to previous releases, a justification is included in the PR description
-   [ ] Tests were added or are not required
-   [ ] Did the new or modified tests pass consistently locally and on CI?
-   [ ] Documentation was added or is not required

### CODE REVIEW

**Comment by shlomi-noach** [comment]
Here is the expected test failure: https://github.com/vitessio/vitess/actions/runs/7736886979/job/21094946752?pr=15103

File: go/vt/vttablet/onlineddl/vrepl/columns_test.go [hunk 1]
```diff
@@ -133,3 +133,248 @@ func TestGetSharedColumns(t *testing.T) {
 		})
 	}
 }
+
```
Comment (shlomi-noach) [inline go/vt/vttablet/onlineddl/vrepl/columns_test.go:136]: This added test is mostly irrelevant to the changes in this PR, but while writing the PR I noticed how this was missing.

File: go/vt/schemadiff/types.go [hunk 2]
```diff
@@ -120,8 +120,8 @@ const (
 )
 
 const (
-	EnumReorderStrategyReject int = iota
-	EnumReorderStrategyAllow
+	EnumReorderStrategyAllow int = iota
```
Comment (shlomi-noach) [inline go/vt/schemadiff/types.go:123]: Subsequently to the main change in this PR, it's now fine to reorder enum values, and the default `schemadiff` hints enum-reordering strategy is now `EnumReorderStrategyAllow`

**Review by dbussink — APPROVED** [review]
Ah, great that tracking the schema is already solved. That makes this much easier!

**Comment by shlomi-noach** [comment]
I believe we should backport this to all supported versions, as this is a bugfix in OnlineDDL/VReplication.

**Review by rohit-nayak-ps — APPROVED** [review]

**Pull request merged**